File: genetic_algorithms/genetic_algorithm.rb
# My Ruby implementation of a basic genetic algorithm
# Based upon:
# Introduction to Genetic Algorithms Chapter 1
# Author: Melanie Mitchell
module NaturalEnvironment
  class << self
    def encapsulate_conditions(args)
      @population_n   = args[:n]   # Total experimental population
      @total_bits     = args[:l]   # Number of coding regions on chromosome
      @generations    = args[:gen] # Generations => # times algorithm runs
      @crossover_rate = args[:pc]  # Chance of chromosomes swap region of bits
      @mutation_rate  = args[:u]   # Chance a given locus switches bits
    end

    # Methods to access the condition variables
    def num_pop
      @population_n
    end

    def num_bits
      @total_bits
    end

    def num_gens
      @generations
    end

    def crossover
      @crossover_rate
    end

    def rate_mute
      @mutation_rate
    end
  end
end

module Nucleus
  class << self
    def recombine(x,y) # Generates new child genomes in the event of crossover
      child_1, child_2 = get_regions(y,x,(rand(y.genes.length) + 1)), get_regions(x,y,(rand(y.genes.length) + 1))
      [Chromosome.new(child_1),Chromosome.new(child_2)]
    end

    def get_regions(genome_1,genome_2,locus) # Takes two xsome ranges and swaps them between the two parents
      genome_1.genes[0, locus] + genome_2.genes[locus, genome_2.genes.length]
    end
  end
end

class Chromosome # Genome and chromosome are used interchangeably due to the nature of the algorithm
  attr_accessor :genes
  def initialize(genes="") # If the genes aren't generated or initialized, randomly create a new xsome
    genes == "" ? self.genes = (1..NaturalEnvironment.num_bits).map { rand(2) }.join : self.genes = genes
  end

  def count
    self.genes.length
  end

  def w               # W denotes fitness in population genetics
    genes.count("1")  # Assuming here that a coding region(bit) w/a value 1  increases fitness
  end                 # By extension 0 are neutral so technically, wrt mutations, slightly deleterious

  def mutate # Randomly mutate genes, changing an individual allele to 1 if it is currently 0
    self.genes.chars.each_with_index do |allele,i|
      allele == "0" ? self.genes[i] = "1" : self.genes[i] = "0" if rand <= NaturalEnvironment.rate_mute
    end
  end
end

class Population
  attr_accessor :chromosomes
  def initialize
    self.chromosomes = []
  end

  def inspect_chromosome # Accesses all the genes of all the xsomes in a given population
    chromosomes.map(&:genes)
  end

  def populate # Populates the population with individuals, represented by xsomes
    self.chromosomes = Array.new(NaturalEnvironment.num_pop) { Chromosome.new }
  end

  def count
    chromosomes.length
  end

  def fitness_values # Accesses the fitness values for each xsome
    chromosomes.map(&:w)
  end

  def w_total
    fitness_values.inject(:+)
  end

  def w_max
    fitness_values.max
  end

  def w_average
    w_total.fdiv(chromosomes.length)
  end

  def select   # Return the selected chromosome from a given population (called twice)
    total = 0
    random_selection = rand(w_total) # Randomly select a value within pop's total fitness
    chromosomes.each_with_index do |xsome, index| # For each xsome in the population
      total += xsome.w                             # Keep a running total of the fitness of each
      (return xsome) if total > random_selection || index == (chromosomes.length - 1)
    end # Return an xsome if the total is above the fitness value or it is the last xsome in the population
  end
end

class Selection
  attr_accessor :prev_gen
  def initialize(args) # Wrap test conditions
    NaturalEnvironment.encapsulate_conditions(args)
  end

  def new_generation_zero # Seed the first generation
    @prev_gen = Population.new
    prev_gen.populate
  end

  def run_generations # Run the test conditions num_gens times
    NaturalEnvironment.num_gens.times do |generation|
      offspring = Population.new # Initialize the new offspring population

      while offspring.count < prev_gen.count # Creates the new generation
        parent_1, parent_2 = select_chromosomes
        child_1, child_2 = analyze_crossover(parent_1,parent_2)
        mutate_chromosomes(child_1,child_2)
        add_offspring(offspring,child_1,child_2)
      end

      # Comment out or remove this method to not display the stats of all generation
      display_selection(generation,prev_gen.w_average.round(2),prev_gen.w_max)
      @prev_gen = offspring # Child generation becomes the new parent generation
    end
  end

  def select_chromosomes
    [prev_gen,prev_gen].map(&:select)
  end

  def analyze_crossover(fn1,fn2) # Randomly chooses whether to recombine parent xsomes
    rand <= NaturalEnvironment.crossover ? Nucleus.recombine(fn1,fn2) : [fn1,fn2]
  end

  def mutate_chromosomes(child_1,child_2)
    [child_1, child_2].map(&:mutate)
  end

  def add_offspring(offspring,child_1,child_2) # Add children: Either both or one, depending upon the test population size
    NaturalEnvironment.num_pop.even? ? offspring.chromosomes.push(child_1,child_2) : offspring.chromosomes.push([child_1, child_2].sample)
  end

  def display_selection(generation,current_w,max_current_w)
    puts "Gen: #{generation} | Avg: #{current_w} | Max: #{max_current_w}"
  end

  def display_stats
    puts ["Generation #{NaturalEnvironment.num_gens}", "Avgerage Fitness: #{prev_gen.w_average.round(2)}", "Max Fitness: #{prev_gen.w_max}"]
  end

  def display_final_xsomes # Formats and displays final xsome bit-strings
    prev_gen.inspect_chromosome.each_with_index { |xsome,i| (i+1) < 10 ? (puts "N#{i+1}:  #{xsome}") : (puts "N#{i+1}: #{xsome}") }
  end
end

drosophila = Selection.new({n: 32, l: 64, gen: 1000, pc: 0.8, u: 0.004})
drosophila.new_generation_zero
drosophila.run_generations
drosophila.display_stats
drosophila.display_final_xsomes

File: genetic_algorithms/genetic_algorithm.exs
defmodule Rand do
  def init do
    << a :: 32, b :: 32, c :: 32 >> = :crypto.strong_rand_bytes(12)
    :random.seed(a,b,c)
  end 

  def num do
    init
    :random.uniform
  end

  def ten_num(len) do
    round(num * len)
  end
end

defmodule World do
  def conditions(init_conditions) do
    Enum.into(init_conditions, Map.new)
  end
end

defmodule Chromosome do
  def w(genes) do
    Regex.scan(~r/1/, genes) |> length()
  end

  def new_base do
    if Rand.num > 0.5, do: "0", else: "1" 
  end

  def set_genes(enviro) do
    1..enviro.num_bits |> Enum.map(fn _ -> new_base end) |> Enum.join()
  end

  def mutate(mutated,gene,enviro,i,len) when i > len, do: mutated
  def mutate(mutated,gene,enviro,i,len) when i <= len do
    if String.at(gene,i) == "0" do
      mutated = "0" <> mutated
    else
      if Rand.num <= enviro.mRate, do: mutated = "0" <> mutated, else: mutated = "1" <> mutated
    end

    mutate(mutated,gene,enviro,i + 1,len)
  end

  def select([xsome|xsomes],  total, rand_select) when total >= rand_select, do: xsome
  def select([xsome|_xsomes], total, rand_select), do: xsome
  def select([xsome|xsomes],  total, rand_select) when total < rand_select do
    select(xsomes,total + Chromosome.w(xsome), rand_select)
  end
end

defmodule Nucleus do
  def recombine(xsome1, xsome2, locus) do
    {child1, child2} = xsome1 |> String.split_at(locus)
    {child3, child4} = xsome2 |> String.split_at(locus)
    child1 <> child4
  end

  def set_xover(xsome1, xsome2) do
    child1 = recombine(xsome1, xsome2, Rand.ten_num(String.length(xsome1)))
    child2 = recombine(xsome2, xsome1, Rand.ten_num(String.length(xsome2)))

    {child1, child2}
  end
end

defmodule Population do
  def populate(pop, i, enviro) when length(pop) >= 32, do: pop
  def populate(pop, i, enviro) when length(pop) <  32  do
    populate(pop ++ [Chromosome.set_genes(enviro)], i + 1, enviro)
  end

  def w([], w_total, w_max, enviro), do: {w_total, w_max}
  def w([xsome|xsomes], w_total, w_max, enviro) do
    fitness = Chromosome.w(xsome)
    if fitness >= w_max, do: w_max = fitness
    w(xsomes, w_total + fitness, w_max, enviro)
  end
end

defmodule Selection do
  def init(enviro) do
    first = Population.populate([], 0, enviro)
    run_generations(first, 0, enviro)
  end

  def run_generations(prev_gen, i, enviro) when i >= 1000 do
    {w_total, w_max} = Population.w(prev_gen, 0, 0, enviro)
    IO.puts "Gen: #{i} | Avg: #{w_total/length(prev_gen)} | Max: #{w_max}"
  end

  def run_generations(prev_gen, i, enviro) when i < 1000 do
    {gen, w_max, w_avg} = calculate_generations(prev_gen, [], enviro)
    IO.puts "Gen: #{i} | Avg: #{w_avg} | Max: #{w_max}"
    run_generations(gen, i + 1, enviro)
  end

  def calculate_generations(prev_gen, offspring, enviro) when length(offspring) >= length(prev_gen) do
    {w_total, w_max} = Population.w(offspring, 0, 0, enviro)
    {offspring, w_max, (w_total/length(offspring))}
  end

  def calculate_generations(prev_gen, offspring, enviro) when length(offspring) < length(prev_gen) do
    {wt, wm} = Population.w(prev_gen, 0, 0, enviro)

    parent1  = Chromosome.select(prev_gen, 0, Rand.ten_num(wt))
    parent2  = Chromosome.select(prev_gen, 0, Rand.ten_num(wt))
    
    {child1, child2} = analyze_crossover(parent1, parent2, enviro)

    child1 = Chromosome.mutate("", child1, enviro, 0, String.length(child1))
    child2 = Chromosome.mutate("", child2, enviro, 0, String.length(child2))

    calculate_generations(prev_gen, [child1, child2] ++ offspring, enviro)
  end

  def analyze_crossover(parent1, parent2, enviro) do
    if Rand.num <= enviro do
      {child1, child2} = Nucleus.set_xover(parent1,parent2)
    else
      {child1, child2} = {parent1, parent2}
    end
    {child1, child2}
  end
end


init_conditions = [n: 32, num_bits: 64, gens: 1000, xoverRate: 0.8, mRate: 0.04]
enviro = World.conditions(init_conditions)
Selection.init(enviro)
